### PR TITLE
Add feature flag for dynamic extension loading (#1878)

### PR DESCRIPTION
**Pull request by mathewc**
Addresses https://github.com/Azure/azure-webjobs-sdk-script/issues/1878

I added a new feature flag to allow dynamic/lazy extension loading to be turned on. FeatureFlag is preferable to a new config knob because feature flags by their very nature are short lived

### CODE REVIEW

File: src/WebJobs.Script/Binding/BuiltinExtensionBindingProvider.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.Azure.WebJobs.Script.Extensibility;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script.Binding
+{
+    /// <summary>
+    /// We have a backwards compat requirement to whitelist #r references to certain "builtin" dlls.
+    /// Hook into #r resolution pipeline and apply the whitelist.
+    /// </summary>
+    internal class BuiltinExtensionBindingProvider : ScriptBindingProvider
```
Comment (mathewc) [inline src/WebJobs.Script/Binding/BuiltinExtensionBindingProvider.cs]: I moved a bunch of this stuff out of ScriptHost into separate Types to improve cohesiveness and code readability

File: src/WebJobs.Script/Binding/ExtensionLoader.cs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+using System.Reflection;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Script.Config;
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Binding
+{
+    internal class ExtensionLoader
```
Comment (mathewc) [inline src/WebJobs.Script/Binding/ExtensionLoader.cs]: All the dynamic extension loading logic is refactored into this new type

File: src/WebJobs.Script/Binding/ExtensionLoader.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                        IExtensionConfigProvider extension = (IExtensionConfigProvider)Activator.CreateInstance(typeExtension);
+                        LoadExtension(extension);
+                    }
+                }
+            }
+        }
+
+        public static IEnumerable<string> DiscoverBindingTypes(IEnumerable<FunctionMetadata> functions)
+        {
+            if (FeatureFlags.IsEnabled(ScriptConstants.FeatureFlagsEnableDynamicExtensionLoading))
```
Comment (mathewc) [inline src/WebJobs.Script/Binding/ExtensionLoader.cs]: This is the change that addresses the issue - we now allow dynamic/lazy loading to be turned on via a feature flag

File: src/WebJobs.Script/Binding/ExtensionLoader.cs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            return false;
+        }
+
+        private void LoadExtension(IExtensionConfigProvider extensionConfigProvider, string locationHint = null)
+        {
+            var extensionConfigProviderType = extensionConfigProvider.GetType();
+            string extensionName = extensionConfigProviderType.Name;
+
+            string msg = null;
+            if (!string.IsNullOrEmpty(locationHint))
```
Comment (mathewc) [inline src/WebJobs.Script/Binding/ExtensionLoader.cs]: Improving the log message. Previously we were getting logs like:

```
```
Loaded custom extension: BotFrameworkConfiguration from ''
Loaded custom extension: SendGridConfiguration from ''
Loaded custom extension: EventGridExtensionConfig from ''
```
```

File: src/WebJobs.Script/Binding/ExtensionLoader.cs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        // The set of "built in" binding types. These are types that are directly accesible without
+        // needing an explicit load gesture.
+        private static IReadOnlyDictionary<string, string> _builtinBindingTypes = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
+        {
+            { "bot", "Microsoft.Azure.WebJobs.Extensions.BotFramework.Config.BotFrameworkConfiguration, Microsoft.Azure.WebJobs.Extensions.BotFramework" },
+            { "sendgrid", "Microsoft.Azure.WebJobs.Extensions.SendGrid.SendGridConfiguration, Microsoft.Azure.WebJobs.Extensions.SendGrid" },
+            { "eventGridTrigger", "Microsoft.Azure.WebJobs.Extensions.EventGrid.EventGridExtensionConfig, Microsoft.Azure.WebJobs.Extensions.EventGrid" }
+        };
+
+        private static IReadOnlyDictionary<string, string> _builtinScriptBindingTypes = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
```
Comment (mamaso) [inline src/WebJobs.Script/Binding/ExtensionLoader.cs]: can we get a comment on the difference between _builtinBindingTypes and _builtinScriptBindingTypes @MikeStall

File: src/WebJobs.Script/Binding/ExtensionLoader.cs (hunk 3 above)
Comment (mamaso) [reply src/WebJobs.Script/Binding/ExtensionLoader.cs]: Is there enough visibility for `AzureWebJobsFeatureFlags` - might need some docs on this

Comment (mathewc) [reply src/WebJobs.Script/Binding/ExtensionLoader.cs]: I added it to https://github.com/Azure/azure-webjobs-sdk-script/wiki/Configuration-Settings

File: src/WebJobs.Script/Binding/ExtensionLoader.cs (hunk 5 above)
Comment (mathewc) [reply src/WebJobs.Script/Binding/ExtensionLoader.cs]: Yeah, wasn't clear to me either

Comment (MikeStall) [reply src/WebJobs.Script/Binding/ExtensionLoader.cs]: _builtinBindingTypes are built on BYOB. This is the correct long-term approach. 
_builtinScriptBindingTypes are not built on BYOB - they have ScriptBindingProviders (which is glue code we aim to get rid of.)

---
In reply to: [138433824](https://github.com/Azure/azure-webjobs-sdk-script/pull/1888#discussion_r138433824) [](ancestors = 138433824)

Comment (mathewc) [reply src/WebJobs.Script/Binding/ExtensionLoader.cs]: I'll add a code comment

**Pull request merged**